File: Combination Sum - Leetcode 39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res, sol = [], []
        nums = candidates
        n = len(nums)

        def backtrack(i, cur_sum):
            if cur_sum == target:
                res.append(sol[:])
                return

            if cur_sum > target or i == n:
                return

            backtrack(i + 1, cur_sum)

            sol.append(nums[i])
            backtrack(i, cur_sum + nums[i])
            sol.pop()

        backtrack(0, 0)
        return res

File: Combination Sum - Leetcode 39/Combination Sum - Leetcode 39.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const res = [];
    const sol = [];
    backtrack(candidates, target, 0, 0, sol, res);
    return res;
};

function backtrack(candidates, target, start, curSum, sol, res) {
    if (curSum === target) {
        res.push([...sol]);
        return;
    }
    if (curSum > target || start === candidates.length) {
        return;
    }

    // Skip the current candidate
    backtrack(candidates, target, start + 1, curSum, sol, res);

    // Include the current candidate
    sol.push(candidates[start]);
    backtrack(candidates, target, start, curSum + candidates[start], sol, res);
    sol.pop();
}
